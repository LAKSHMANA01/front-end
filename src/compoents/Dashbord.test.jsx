// Dashboard.test.jsx
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Dashboard from './Dashbord';

// Dummy data for the Dashboard cards
const dummyData = {
  openTickets: [1, 2, 3],
  allTasks: [1, 2],
  failedTasks: [1],
  inprogressTasks: [1, 2, 3, 4],
  allEngineers: [1, 2, 3, 4],
  resolvedTickets: [1, 2],
  approvedEngineers: [1, 2],
  completedTasks: [1],
  pendingTasks: [],
  totalTasks: [1, 2, 3, 4, 5],
  activeEngineers: [],
  allHazards: [1, 2, 3],
};

// Valid and invalid chart data for the Bar charts
const ticketStatusDataValid = { labels: ['A', 'B'], datasets: [] };
const taskPriorityDataValid = { labels: ['X', 'Y'], datasets: [] };
const ticketStatusDataInvalid = {}; // missing labels
const taskPriorityDataInvalid = {};

// --- Mocks ---

// Mock the Card component so we can inspect the props passed by Dashboard.
jest.mock('./Card', () => (props) => (
  <div data-testid="card" data-title={props.title} data-value={props.value}>
    {props.icon}
  </div>
));

// Mock the Bar chart component. It will render a div with text "Chart" if valid data is provided.
jest.mock('react-chartjs-2', () => ({
  Bar: (props) => (
    <div data-testid="bar-chart" data-data={JSON.stringify(props.data)}>
      {props.data && props.data.labels ? 'Chart' : null}
    </div>
  ),
  Pie: () => <div>Pie</div>,
}));

// Mock ToastContainer to verify it is rendered.
jest.mock('react-toastify', () => ({
  ToastContainer: (props) => <div data-testid="toast-container" {...props} />,
}));

// --- Test Suite ---
describe('Dashboard Component', () => {
  // Clear sessionStorage before each test so that each test can set its own role.
  beforeEach(() => {
    sessionStorage.clear();
  });

  // Helper function to render the Dashboard with desired props and role.
  const renderDashboard = (role, ticketStatusData, taskPriorityData) => {
    sessionStorage.setItem('role', role);
    return render(
      <Dashboard
        ticketStatusData={ticketStatusData}
        taskPriorityData={taskPriorityData}
        data={dummyData}
      />
    );
  };

  test('renders the header "Dashboard"', () => {
    renderDashboard('user', ticketStatusDataValid, taskPriorityDataValid);
    expect(screen.getByText('Dashboard')).toBeInTheDocument();
  });

  test('renders user role cards correctly', () => {
    renderDashboard('user', ticketStatusDataValid, taskPriorityDataValid);
    const cards = screen.getAllByTestId('card');
    expect(cards.length).toBe(4);

    // For "user" role, the card configuration should be:
    // - Total Tickets: totalTasks.length === 5
    // - Completed Tickets: completedTasks.length === 1
    // - Pending Tickets: totalTasks.length + inprogressTasks.length === 5 + 4 = 9
    // - Failed Tickets: failedTasks.length === 1
    expect(cards[0]).toHaveAttribute('data-title', 'Total Tickets');
    expect(cards[0]).toHaveAttribute('data-value', '5');

    expect(cards[1]).toHaveAttribute('data-title', 'Completed Tickets');
    expect(cards[1]).toHaveAttribute('data-value', '1');

    expect(cards[2]).toHaveAttribute('data-title', 'Pending Tickets');
    expect(cards[2]).toHaveAttribute('data-value', '9');

    expect(cards[3]).toHaveAttribute('data-title', 'Failed Tickets');
    expect(cards[3]).toHaveAttribute('data-value', '1');
  });

  test('renders engineer role cards correctly', () => {
    renderDashboard('engineer', ticketStatusDataValid, taskPriorityDataValid);
    const cards = screen.getAllByTestId('card');
    expect(cards.length).toBe(4);

    // For "engineer" role, the configuration should be:
    // - My Active Tickets: allTasks.length === 2
    // - Pending Tickets: allTasks.length + inprogressTasks.length === 2 + 4 = 6
    // - Completed Tickets: resolvedTickets.length === 2
    // - All Hazards: allHazards.length === 3
    expect(cards[0]).toHaveAttribute('data-title', 'My Active Tickets');
    expect(cards[0]).toHaveAttribute('data-value', '2');

    expect(cards[1]).toHaveAttribute('data-title', 'Pending Tickets');
    expect(cards[1]).toHaveAttribute('data-value', '6');

    expect(cards[2]).toHaveAttribute('data-title', 'Completed Tickets');
    expect(cards[2]).toHaveAttribute('data-value', '2');

    expect(cards[3]).toHaveAttribute('data-title', 'All Hazards');
    expect(cards[3]).toHaveAttribute('data-value', '3');
  });

  test('renders admin role cards correctly', () => {
    renderDashboard('admin', ticketStatusDataValid, taskPriorityDataValid);
    const cards = screen.getAllByTestId('card');
    expect(cards.length).toBe(4);

    // For "admin" role, the configuration should be:
    // - Open Tickets: openTickets.length === 3
    // - Resolved Tickets: resolvedTickets.length === 2
    // - New Engineers: allEngineers.length - approvedEngineers.length === 4 - 2 = 2
    // - Approved Engineers: approvedEngineers.length === 2
    expect(cards[0]).toHaveAttribute('data-title', 'Open Tickets');
    expect(cards[0]).toHaveAttribute('data-value', '3');

    expect(cards[1]).toHaveAttribute('data-title', 'Resolved Tickets');
    expect(cards[1]).toHaveAttribute('data-value', '2');

    expect(cards[2]).toHaveAttribute('data-title', 'New Engineers');
    expect(cards[2]).toHaveAttribute('data-value', '2');

    expect(cards[3]).toHaveAttribute('data-title', 'Approved Engineers');
    expect(cards[3]).toHaveAttribute('data-value', '2');
  });

  test('renders "No data available" when role is not recognized', () => {
    renderDashboard('other', ticketStatusDataValid, taskPriorityDataValid);
    expect(screen.getByText('No data available')).toBeInTheDocument();
  });

  test('renders valid Bar charts when data is provided', () => {
    renderDashboard('user', ticketStatusDataValid, taskPriorityDataValid);
    const barCharts = screen.getAllByTestId('bar-chart');
    // Two charts should be rendered (one for ticket status and one for task priority)
    expect(barCharts.length).toBe(2);
    barCharts.forEach((chart) => {
      expect(chart).toHaveTextContent('Chart');
    });
  });

  test('renders fallback text for ticketStatusData and taskPriorityData when labels are missing', () => {
    renderDashboard('user', ticketStatusDataInvalid, taskPriorityDataInvalid);
    // When labels are missing, fallback texts should be rendered.
    expect(screen.getByText('No Ticket Data')).toBeInTheDocument();
    expect(screen.getByText('No Task Priority Data')).toBeInTheDocument();
  });

  test('renders the ToastContainer', () => {
    renderDashboard('user', ticketStatusDataValid, taskPriorityDataValid);
    expect(screen.getByTestId('toast-container')).toBeInTheDocument();
  });

  test('matches the snapshot', () => {
    const { container } = renderDashboard('admin', ticketStatusDataValid, taskPriorityDataValid);
    expect(container).toMatchSnapshot();
  });
});