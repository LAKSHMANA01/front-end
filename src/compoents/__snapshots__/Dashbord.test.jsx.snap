// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component matches the snapshot 1`] = `
<div>
  <div
    class="grow p-6 dark:bg-gray-900 dark:text-white"
  >
    <h1
      class=" font-bold bg-white rounded-md  justify-start text-2xl  w-full h-50  p-3 mb-6"
    >
       Dashboard  
    </h1>
    <div
      class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-4"
    >
      <div
        data-testid="card"
        data-title="Open Tickets"
        data-value="3"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 384 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
          />
        </svg>
      </div>
      <div
        data-testid="card"
        data-title="Resolved Tickets"
        data-value="2"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 384 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M336 64h-80c0-35.3-28.7-64-64-64s-64 28.7-64 64H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48V112c0-26.5-21.5-48-48-48zM192 40c13.3 0 24 10.7 24 24s-10.7 24-24 24-24-10.7-24-24 10.7-24 24-24zm121.2 231.8l-143 141.8c-4.7 4.7-12.3 4.6-17-.1l-82.6-83.3c-4.7-4.7-4.6-12.3.1-17L99.1 285c4.7-4.7 12.3-4.6 17 .1l46 46.4 106-105.2c4.7-4.7 12.3-4.6 17 .1l28.2 28.4c4.7 4.8 4.6 12.3-.1 17z"
          />
        </svg>
      </div>
      <div
        data-testid="card"
        data-title="New Engineers"
        data-value="2"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M624 208h-64v-64c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v64h-64c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h64v64c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-64h64c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm-400 48c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
          />
        </svg>
      </div>
      <div
        data-testid="card"
        data-title="Approved Engineers"
        data-value="2"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 224c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm448 0c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm32 32h-64c-17.6 0-33.5 7.1-45.1 18.6 40.3 22.1 68.9 62 75.1 109.4h66c17.7 0 32-14.3 32-32v-32c0-35.3-28.7-64-64-64zm-256 0c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C179.6 288 128 339.6 128 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zm-223.7-13.4C161.5 263.1 145.6 256 128 256H64c-35.3 0-64 28.7-64 64v32c0 17.7 14.3 32 32 32h65.9c6.3-47.4 34.9-87.3 75.2-109.4z"
          />
        </svg>
      </div>
    </div>
    <div
      class="grid grid-cols-1 md:grid-cols-2 gap-6 mt-6"
    >
      <div
        class="bg-white p-4 rounded-lg shadow-md dark:bg-gray-800"
      >
        <h3
          class="text-lg font-semibold mb-4"
        >
          Ticket Status Overview
        </h3>
        <div
          data-data="{"labels":["A","B"],"datasets":[]}"
          data-testid="bar-chart"
        >
          Chart
        </div>
      </div>
      <div
        class="bg-white p-4 rounded-lg shadow-md dark:bg-gray-800"
      >
        <h3
          class="text-lg font-semibold mb-4"
        >
          Task Priority Overview
        </h3>
        <div
          data-data="{"labels":["X","Y"],"datasets":[]}"
          data-testid="bar-chart"
        >
          Chart
        </div>
      </div>
    </div>
    <div
      autoclose="5000"
      data-testid="toast-container"
      draggable="true"
      position="top-right"
    />
  </div>
</div>
`;
